tests/test_io.py

```diff
@@ -277,6 +277,15 @@ def test_list_dir_names_errors(tmpdir):
             False,
             False,
         ),
+        pytest.param(
+            [os.path.join('sub', 'file.txt')],
+            'file.txt',
+            '',
+            [],
+            False,
+            False,
+            marks=pytest.mark.xfail(raises=NotADirectoryError),
```

Comment:
Can you repeat this test with `recursive=True` there it should not fail, but return the file pattern.

Reply:
As I pointed out before, it will still fail. Since `file.txt` does not exist and it is not pattern, it will treat it as directory and since the directory does not exist, it raises an error.

Reply:
I was also expecting it to fail, but I would argue that is not the behavior a user would expect.
Because now what `list_file_names()` returns for the sub folder does not depend on the actual content of the sub-folder, but on the content of the main folder, which feels very strange.

Reply:
I know, but I don't see a solution to this as long as we don't introduce a separate `pattern` argument.

So maybe we should only support `recursive=True` when `path` is a folder or ends on a pattern. I.e.  `list_file_names('./file.txt', recursive=True)` will either fail if `'./file.txt'` does not exist or return `['./file.txt']` (even if e.g. `'./sub/file.txt'` exists).

Reply:
There might also be the possibility to implement a solution to return `file.txt` from a sub-folder.
E.g. we could not only check `os.path.exists(path)`, but we could do it for every sub-folder if `recursive` is `True`.

Reply:
That's indeed a solution. I implemented your suggestion and now the following works as expected:

```python
>>> audeer.mkdir('dir')
>>> audeer.mkdir('dir/sub')
>>> audeer.touch('dir/sub/file.txt')
>>> audeer.list_file_names('file.txt', basenames=True, recursive=True))
['dir/sub/file.txt']
```

There's one more special case that we now treat as follows:

```python
>>> audeer.touch('.file.txt')
>>> audeer.list_file_names(.file.txt', basenames=True)
[]
>>> audeer.list_file_names(.file.txt', basenames=True, hidden=True)
['.file.txt']
```

Or would you expect it raises an error in the first case?

Reply:
No, I think this is a good solution.

It also works for a combination of hidden files in sub-folders:

```python
>>> audeer.mkdir(sub')
>>> audeer.touch('sub/.file.txt')
>>> audeer.list_file_names('.file.txt', basenames=True, recursive=True, hidden=True)
['sub/.file.txt']
>>> audeer.list_file_names('.file.txt', basenames=True, hidden=True)
...
NotADirectoryError: /home/audeering.local/hwierstorf/git/audeering/audeer/.file.txt
```

But there is one case where it is not working yet:

```python
>>> # the next should return []
>>> audeer.list_file_names('.file.txt', basenames=True, recursive=True)
...
NotADirectoryError: /home/audeering.local/hwierstorf/git/audeering/audeer/.file.txt
```

Reply:
Good catch. This case is now covered and I added a test for it.

Reply:
Cool, thanks.